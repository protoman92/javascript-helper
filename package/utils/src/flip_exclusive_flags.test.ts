import { flipMutualExclusiveFlags } from ".";

describe("Flip exclusive flags", () => {
  it("Should work correctly", async () => {
    // Setup
    const allFlags = { a: true, b: true, c: true };
    const mappers = { a: () => 1, b: () => 2, c: () => 3 };
    const flipperWithMappers = flipMutualExclusiveFlags({ mappers });

    // When
    expect(
      flipMutualExclusiveFlags({
        allFlags,
        flagToSetTrue: "a",
      })
    ).toMatchSnapshot();

    expect(
      flipMutualExclusiveFlags({
        allFlags,
        flagToSetTrue: "b",
      })
    ).toMatchSnapshot();

    expect(
      flipMutualExclusiveFlags({
        allFlags,
        mappers,
        flagToSetTrue: "a",
      })
    ).toMatchSnapshot();

    expect(
      flipMutualExclusiveFlags({
        allFlags,
        mappers,
        flagToSetTrue: "b",
      })
    ).toMatchSnapshot();

    expect(
      flipperWithMappers({
        allFlags,
        flagToSetTrue: "a",
      })
    ).toMatchSnapshot();

    expect(
      flipperWithMappers({
        allFlags,
        flagToSetTrue: "b",
      })
    ).toMatchSnapshot();
  });

  it("Flipper with pre-supplied mappers should maintain type-safety", async () => {
    // Setup
    const flipper = flipMutualExclusiveFlags({
      mappers: { a: (f) => !f, b: (f) => !f },
    });

    // When && Then
    expect(
      flipper({ allFlags: { a: true, b: true }, flagToSetTrue: "a" })
    ).toMatchSnapshot();

    expect(
      flipper({ allFlags: { a: true, b: true }, flagToSetTrue: "b" })
    ).toMatchSnapshot();

    expect(
      flipper({
        allFlags: { a: true, b: true, c: true } as any,
        flagToSetTrue: "a",
      })
    ).toMatchSnapshot();
  });
});
