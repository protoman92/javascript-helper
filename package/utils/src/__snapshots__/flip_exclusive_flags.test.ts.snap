// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flip exclusive flags Flipper with pre-supplied mappers should maintain type-safety 1`] = `
{
  "a": false,
  "b": true,
}
`;

exports[`Flip exclusive flags Flipper with pre-supplied mappers should maintain type-safety 2`] = `
{
  "a": true,
  "b": false,
}
`;

exports[`Flip exclusive flags Flipper with pre-supplied mappers should maintain type-safety 3`] = `
{
  "a": false,
  "b": true,
}
`;

exports[`Flip exclusive flags Should work correctly 1`] = `
{
  "a": true,
  "b": false,
  "c": false,
}
`;

exports[`Flip exclusive flags Should work correctly 2`] = `
{
  "a": false,
  "b": true,
  "c": false,
}
`;

exports[`Flip exclusive flags Should work correctly 3`] = `
{
  "a": 1,
  "b": 2,
  "c": 3,
}
`;

exports[`Flip exclusive flags Should work correctly 4`] = `
{
  "a": 1,
  "b": 2,
  "c": 3,
}
`;

exports[`Flip exclusive flags Should work correctly 5`] = `
{
  "a": 1,
  "b": 2,
  "c": 3,
}
`;

exports[`Flip exclusive flags Should work correctly 6`] = `
{
  "a": 1,
  "b": 2,
  "c": 3,
}
`;
